### PR TITLE
Fix hook parser

### PR DESCRIPTION
**Pull request by strickvl**
Noticed an error when running the hooks example (also on the video). Looked into it and seems that we're passing a `str` type as the `arg_type` always and so were always hitting the `TypeError` exception. Not sure if this is *the* solution, but proposing it as a possible solution.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/orchestrators/step_runner.py [hunk 1]
```diff
@@ -315,6 +315,11 @@ def _parse_hook_inputs(
             arg_type = annotations.get(arg, None)
             arg_type = resolve_type_annotation(arg_type)
 
+            if isinstance(arg_type, str):
```
Comment (schustmi) [inline src/zenml/orchestrators/step_runner.py:318]: We don't allow string annotations in any other place for probably the same reason, is there any specific reason we require them for hooks?

This lookup in globals/locals seems pretty unreliable to me, what if a the class that is annotated is not in one of those dicts?

**Comment by strickvl** [comment]
> Thanks for catching this! Could you let me know what the error is exactly (Is it an exception or a warning or something?).

ScreenShot 2023-03-21 at 12 38 43

@htahir1 This is the message that is output when you run it without the fix.

@schustmi I didn't dive deep into the hooks code so not sure where the string annotations come from, but I'll take a look and see if we can just improve it at the source.

**Comment by htahir1** [comment]
So the hooks never worked at all if i understood this correctly?

**Comment by strickvl** [comment]
> So the hooks never worked at all if i understood this correctly?

I think not.

File: src/zenml/orchestrators/step_runner.py [hunk 2]
```diff
@@ -531,8 +531,9 @@ def load_and_run_hook(
             output_materializers: The output materializers of the step.
         """
         try:
-            hook = source_utils.load_source_path(hook_source)
+            hook = source_utils.import_class_by_path(hook_source)
```
Comment (schustmi) [inline src/zenml/orchestrators/step_runner.py:534]: Just out of curiosity, can you explain why this change necessary here?

**Review by schustmi — APPROVED** [review]

File: src/zenml/orchestrators/step_runner.py (hunk 2 above)
Comment (strickvl) [reply src/zenml/orchestrators/step_runner.py:534]: Good catch. Actually it might not be any more. I can probably just revert it back. I changed this early on when we were grabbing objects from globals and locals. I'll revert it and rerun the tests.

Comment (strickvl) [reply src/zenml/orchestrators/step_runner.py:534]: Indeed the change was unnecessary with the later fix we made.

File: src/zenml/hooks/hook_validators.py [hunk 3]
```diff
@@ -67,10 +67,6 @@ def resolve_and_validate_hook(hook_func: "HookSpecification") -> str:
                     BaseException,
                     BaseParameters,
                     StepContext,
-                    # Have to do string version for TYPE_CHECKING
-                    "BaseException",
```
Comment (htahir1) [inline src/zenml/hooks/hook_validators.py:71]: But I added these just for the case.. Im very unsure about why we revert this now but if @schustmi and @strickvl are confident id say merge this PR in...

Comment (schustmi) [reply src/zenml/hooks/hook_validators.py:71]: That case doesn't really work in a nice way though: What if I pass my custom `BaseParameters` subclass as the type annotation there, it would already break down. I think we should just not allow these for now until we implement an actual solution to support string annotations in our steps/hooks.

**Review by htahir1 — APPROVED** [review]
